pydicom/fileset.py

```diff
@@ -0,0 +1,2720 @@
+
+        Parameters
+        ----------
+        ds_or_path : pydicom.dataset.Dataset, str or PathLike
+            An existing File-set's DICOMDIR, either as a pydicom
+            :class:`~pydicom.dataset.Dataset` or the path to the DICOMDIR file
+            as :class:`str` or :class:`pathlib.Path`.
+        include_orphans : bool, optional
+            If ``True`` (default) include instances referenced by orphaned
+            directory records in the File-set.
+        raise_orphans : bool, optional
+            If ``True`` then raise an exception if orphaned directory records
+            are found in the File-set (default ``False``).
+        """
+        if isinstance(ds_or_path, Dataset):
+            ds = ds_or_path
+        else:
+            ds = dcmread(ds_or_path)
+
+        sop_class = ds.file_meta.get("MediaStorageSOPClassUID", None)
+        if sop_class != MediaStorageDirectoryStorage:
+            raise ValueError(
+                "Unable to load the File-set as the supplied dataset is "
+                "not a 'Media Storage Directory' instance"
+            )
+
+        tsyntax = ds.file_meta.TransferSyntaxUID
+        if tsyntax != ExplicitVRLittleEndian:
+            warnings.warn(
+                "The DICOMDIR dataset uses an invalid transfer syntax "
+                f"'{tsyntax.name}' and will be updated to use 'Explicit VR "
+                "Little Endian'"
+            )
+
+        try:
+            path = Path(ds.filename).resolve(strict=True)
+        except FileNotFoundError:
+            raise FileNotFoundError(
+                "Unable to load the File-set as the 'filename' attribute "
+                "for the DICOMDIR dataset is not a valid path: "
+                f"{ds.filename}"
+            )
+        except TypeError:
+            # Custom message if DICOMDIR from bytes, etc
+            raise TypeError(
+                "Unable to load the File-set as the DICOMDIR dataset must "
+                "have a 'filename' attribute set to the path of the "
+                "DICOMDIR file"
+            )
+
+        self.clear()
+        self._id = ds.get("FileSetID", None)
+        self._uid = ds.file_meta.get(
+            "MediaStorageSOPInstanceUID", generate_uid()
+        )
+        self._descriptor = ds.get("FileSetDescriptorFileID", None)
+        self._charset = ds.get(
+            "SpecificCharacterSetOfFileSetDescriptorFile", None
+        )
+        self._path = path.parent
+        self._ds = ds
+
+        # Create the record tree
+        self._parse_records(ds, include_orphans, raise_orphans)
+
+        bad_instances = []
+        for instance in self:
+            # Check that the referenced file exists
+            file_id = instance.node._file_id
+            if file_id is None:
+                bad_instances.append(instance)
+                continue
+
+            (self.path / file_id).resolve(strict=True)
+            # If the instance's existing directory structure doesn't match
```

Comment:
Is the resolved path return value not wanted, or is it done just to throw a potential error?

Reply:
Just to throw the error

Reply:
It might be better to warn instead and remove the instance.

Reply:
Done